### ISSUE TITLE
Poor data loading performance and memory leaks

### ISSUE THREAD

**Issue by xiazhongyv** [aharley/simple_bev#4]
During training, every time my GPU worked for 5.5 seconds, it had to wait 11 seconds for the Dataloader. I wonder if you or anyone else have encountered this problem. 
I think this code may cause the Dataloader to get items more slowly:
Loops (nuscnensdataset.py L1130, L1174-1175), 
Lots of list.appends and torch.stacks, 
Redundant Tensor.clone operations(nuscnensdataset.py L990,996,997), 
A lot of CPU (numpy) operations, rather than loading them onto the GPU and then using GPU (torch.Tensor) operations.
This could be a fatal problem, given the weak CPU of on-board devices such as NVIDIA Orin.

Besides, I have frequently encountered problems that child processes were not properly reclaimed when training program terminated, resulting in memory leaks.

Please check. Thank you for your work.

**Comment by aharley** [comment]
This sounds great, thanks for bringing this up. In general, I have not actually seen the dataloader be a bottleneck in the current code. Usually `rtime` goes close to 0 after the first few iterations. But I often use a large `nworkers` argument and maybe I'm just overpowering the issue. 

I'm a bit confused by the suggestion to put operations on the GPU -- in my experience, it's best if all preprocessing happens in the dataloader, and it's best if everything everything in the dataloader happens on cpu. 

For the memory leaks: can you provide more info? Maybe I can fix it.

I can reduce the number of clones easily. Eliminating the loops you pointed to is much harder.

**Comment by xiazhongyv** [comment]
Sorry, I didn't make it clear. I mean it might be better to put some operations outside of Dataloader, like fetching the last frame from img.

For the memory leaks: 
When using CTRL + C to terminate a program, or when a program ends normally after all steps were finished, there is a certain chance that child processes will not be destroyed (or new child processes be created ? ). The GPU memory is full, and the GPU usage is 100%. 
I'm not quite sure how to diagnose this problem.

**Comment by aharley** [comment]
I made some partial progress, by reducing/rewriting the big `clone()` lines you found. Seems like only a negligible effect on my side. By the way, if you pull again, the code might ask you to install the Lyft SDK into your environment (`pip install -U git+https://github.com/lyft/nuscenes-devkit`).

> like fetching the last frame from `img`
Can you be more specific here? I'm not sure what this refers to.

Thanks for describing the memory leak more. Right now I'm thinking: I would have certainly noticed a leak such as that, so I believe it's not happening for me. For example, whenever a run ends for me (either killed or ended normally), the memory on all GPUs is 0MB. If you can find out more, please do write about it.

**Comment by GANWANSHUI** [comment]
> 8x5_3e-4s_rgb_mine_00:59:30; step 000678/10000; rtime 7.22; itime 10.03; loss 3.90437; iou_t 29.3; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000679/10000; rtime 0.33; itime 3.14; loss 4.27795; iou_t 30.3; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000680/10000; rtime 0.57; itime 3.32; loss 3.07320; iou_t 30.4; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000681/10000; rtime 9.18; itime 11.93; loss 3.04413; iou_t 31.8; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000682/10000; rtime 0.39; itime 2.88; loss 3.75056; iou_t 31.9; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000683/10000; rtime 6.87; itime 10.10; loss 3.75146; iou_t 31.9; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000684/10000; rtime 0.36; itime 3.22; loss 4.42175; iou_t 32.2; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000685/10000; rtime 3.25; itime 6.27; loss 3.52828; iou_t 33.7; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000686/10000; rtime 4.19; itime 6.84; loss 5.39904; iou_t 34.4; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000687/10000; rtime 0.22; itime 3.43; loss 3.89521; iou_t 33.5; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000688/10000; rtime 8.14; itime 10.96; loss 3.69717; iou_t 32.4; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000689/10000; rtime 0.29; itime 3.17; loss 4.47343; iou_t 32.4; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000690/10000; rtime 5.03; itime 8.00; loss 3.20230; iou_t 33.0; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000691/10000; rtime 4.30; itime 6.64; loss 3.31642; iou_t 32.1; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000692/10000; rtime 0.20; itime 3.19; loss 3.70647; iou_t 31.9; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000693/10000; rtime 5.86; itime 8.66; loss 3.57874; iou_t 32.8; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000694/10000; rtime 0.34; itime 3.11; loss 3.63707; iou_t 33.5; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000695/10000; rtime 9.03; itime 11.94; loss 3.46396; iou_t 32.4; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000696/10000; rtime 0.25; itime 3.27; loss 3.13473; iou_t 31.4; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000697/10000; rtime 7.28; itime 9.99; loss 2.98346; iou_t 32.6; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000698/10000; rtime 0.55; itime 2.97; loss 3.01132; iou_t 33.8; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000699/10000; rtime 0.29; itime 2.99; loss 4.04172; iou_t 33.5; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000700/10000; rtime 8.00; itime 12.85; loss 4.23140; iou_t 32.8; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000701/10000; rtime 0.15; itime 2.31; loss 4.70462; iou_t 32.9; iou_v nan
8x5_3e-4s_rgb_mine_00:59:30; step 000702/10000; rtime 1.97; itime 4.00; loss 3.75331; iou_t 32.8; iou_v nan

Hi， I also have some trouble with the training. Somehow the training just stacks on some iterations and we can see above that the rtime and itime are quite inconsistent in different iterations. I used the A100 40GB to train the network in the default setting. Any idea for this situation? Thank you very much!

**Comment by aharley** [comment]
@GANWANSHUI have you tried increasing `nworkers`? You may want to max this out for your system. Also, if you use a faster drive then this problem may disappear. 

The inconsistent readtime is just what it looks like when the dataloader needs to "catch up". 

I'll try to push a speed up this weekend...

**Comment by GANWANSHUI** [comment]
Thanks a lot for the response. I try to increase the nworkers to 24 and the rtime and itime are normal and consistent. But there is still a problem that the training will stuck in some iteration and it seems the suspension is random. Thanks a lot if having some tips on this!

**Comment by aharley** [comment]
@GANWANSHUI I pushed a speedup. Data loading is 2x faster for me now. Can you try it out?

**Comment by GANWANSHUI** [comment]
@aharley  Thanks a lot, I will try it recently and give feedback.  BTW, from your code base, I find the different view transformers: Bevformer and the liftnet2. May I ask that does the code can function? If yes, I will try with them too.  Thanks again!

**Comment by aharley** [comment]
Those are just experimental, so I don't recommend them really.
